src/lib/optimizer/abstract_syntax_tree/abstract_ast_node.cpp

```diff
@@ -181,37 +224,56 @@ std::vector<ColumnID> AbstractASTNode::get_output_column_ids_for_table(const std
     return get_output_column_ids();
   }
 
-  return _left_child->get_output_column_ids_for_table(table_name);
+  return left_child()->get_output_column_ids_for_table(table_name);
 }
 
 void AbstractASTNode::remove_from_tree() {
-  Assert(!_right_child, "Can't remove a node with two children");
+  Assert(left_child() && !right_child(), "Can only rermove nodes that only have a left child");
 
-  auto parent = _parent.lock();
+  /**
+   * Back up parents and in which child side they hold this node
+   */
+  auto parents = this->parents();
+  auto child_sides = this->get_child_sides();
 
-  if (parent) {
-    parent->set_left_child(_left_child);  // Note: It's totally fine for _left_child to be a nullptr
-  } else if (_left_child) {
-    _left_child->clear_parent();
+  // This will call get_child_side() again, but let's avoid code duplication
+  clear_parents();
+
+  /**
+   * Hold left_child ptr in extra variable and untie it from this node
+   */
+  auto left_child = this->left_child();
+  set_left_child(nullptr);
+
+  /**
+   * Tie this nodes previous parents with this nodes previous left child
+   */
+  if (left_child) {
+    for (size_t parent_idx = 0; parent_idx < parents.size(); ++parent_idx) {
+      parents[parent_idx]->set_child(child_sides[parent_idx], left_child);
+    }
   }
 }
 
 void AbstractASTNode::replace_in_tree(const std::shared_ptr<AbstractASTNode>& node_to_replace) {
```

Comment:
Naming: I think the direction is somewhat unclear. Wouldn't `node_to_replace.replace_by` be clearer?